### PR TITLE
Do not simplify VB ObjectCreationExpression initializers into inferred initializers

### LINKED ISSUE #43475: Simplifying VB ObjectCreationExpression creates compile error when inferred name matches
**Version Used**: d84800ba526f3cf9ffcb3eed0e88a6219beab5d9

**Steps to Reproduce**:

Run simplifier on this VB code:
```vbnet
New StringBuilder With { .Length = sb.Length }
```

**Expected Behavior**:
```vbnet
New StringBuilder With { .Length = sb.Length }
```

**Actual Behavior**:
```vbnet
New StringBuilder With { sb.Length }
```
This is not valid syntax for an object creation expression, but would be for an anonymous type etc. which is what the simplifier was clearly targeting. 

**Other details**:
A similar fix was made to disable the diagnostic in VS. It's a pity they don't share this code: https://github.com/dotnet/roslyn/pull/23817/files#diff-aad522f9b1eb080adef996187a2a1b28R65

### PR DESCRIPTION
**Pull request by GrahamTheCoder**
Fixes #43475

* The test TestVisualBasic_DontRemovePropertyNameForObjectCreationInitializer initially failed because the binder threw an exception.
  *  In fact, if you're within Visual Studio, the FailFast error handler kicks in and kills Visual Studio. I've raised that as an issue separately: https://github.com/dotnet/roslyn/issues/41724
* The other test is just to show that the simplification still works correctly in a related case I didn't change in this PR.

### Out of scope (separate issue)

The first place that blew up with the invalid syntax tree created by the simplifier was the binder. I haven't done anything to change that behaviour.

In my ideal world it shouldn't get as far as the binder - either the SyntaxFactory method, or `new ObjectCreationExpressionSyntax` should throw an exception if creating an `ObjectCreationInitializerSyntax` which either:
* Isn't a `ObjectMemberInitializerSyntax`
* Has an initializer which is not a `NamedFieldInitializerSyntax`

It's not immediately clear if this is possible/easy since the file is generated from xml. If someone gives me a pointer to somewhere similar logic has been implemented, I'll happily address this in a separate PR.

### CODE REVIEW

**Comment by GrahamTheCoder** [comment]
In my tweaks to the test base I made it possible to easily check for compile errors in the output. Could be used for https://github.com/dotnet/roslyn/pull/43124 and a couple of other similar cases I hope to get to once these are done.

**Comment by GrahamTheCoder** [comment]
Conflicts fixed, ready to merge

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Simplification/VisualBasicInferredMemberNameSimplifier.vb [hunk 1]
```diff
@@ -32,7 +32,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Simplification
         End Function
 
         Friend Function CanSimplifyNamedFieldInitializer(node As NamedFieldInitializerSyntax) As Boolean
-            If RemovalCausesAmbiguity(DirectCast(node.Parent, ObjectMemberInitializerSyntax).Initializers, node) Then
+            If node.Parent.IsParentKind(SyntaxKind.ObjectCreationExpression) OrElse RemovalCausesAmbiguity(DirectCast(node.Parent, ObjectMemberInitializerSyntax).Initializers, node) Then
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Simplification/VisualBasicInferredMemberNameSimplifier.vb:35]: 1. should likely have omment explaining this. 
2. what happens with nested initializers?  you can have those in C# like so:

```c#
new Something
{
    C = { Y = y, Z  = z },
}
```

woudl that be an issue for VB here?

Comment (GrahamTheCoder) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Simplification/VisualBasicInferredMemberNameSimplifier.vb]: Yep this is a good point. The previous code was more robust in that it enforced with an `InvalidCastException` that the parent was an `ObjectMemberInitializer`. I've pulled that out as a variable to maintain the old behaviour. That way there'll be no jumping through an unknown intermediate node and no risk of what you've described.

I have also now checked this is correct according to [syntax in the docs](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/language-specification/expressions#object-creation-expressions) which say any `FieldInitializer` is always part of an `ObjectMemberInitializer`. They also say `ObjectMemberInitializer` is always part of `AnonymousObjectCreationExpression` or `ObjectCreationExpressionInitializer` (they're from 2017 so don't mention named tuples).

I felt a bit thin on information to add to the comment so if there's any questions you have that you feel a comment should explain let me know

File: src/EditorFeatures/Test2/Simplification/InitializerSimplificationTests.vb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                Dim semanticModel = Await simplifiedDocument.GetSemanticModelAsync()
+                Dim diagnosticsFromSimplifiedDocument = semanticModel.Compilation.GetDiagnostics()
+                Assert.Empty(diagnosticsFromSimplifiedDocument)
+
+                Await AssertCodeEqual(expected, simplifiedDocument)
+            End Using
+        End Function
+    End Class
+
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Simplification/InitializerSimplificationTests.vb:100]:
```suggestion
```

File: src/EditorFeatures/Test2/Simplification/InitializerSimplificationTests.vb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    End Function
+End Class
+</code>
+            Await AssertCompilesAndEqual(input, expected).ConfigureAwait(False)
+        End Function
+#End Region
+
+
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Simplification/InitializerSimplificationTests.vb:86]:
```suggestion

```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Simplification/VisualBasicInferredMemberNameSimplifier.vb [hunk 4]
```diff
@@ -32,7 +32,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Simplification
         End Function
 
         Friend Function CanSimplifyNamedFieldInitializer(node As NamedFieldInitializerSyntax) As Boolean
-            If RemovalCausesAmbiguity(DirectCast(node.Parent, ObjectMemberInitializerSyntax).Initializers, node) Then
+            Dim parentMemberInitializer As ObjectMemberInitializerSyntax = DirectCast(node.Parent, ObjectMemberInitializerSyntax)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Simplification/VisualBasicInferredMemberNameSimplifier.vb:35]:
```suggestion
            Dim parentMemberInitializer = DirectCast(node.Parent, ObjectMemberInitializerSyntax)
```

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**